### PR TITLE
Typesafe (kinda) Attributes now available

### PR DESCRIPTION
**Pull request by chewxy**
So, the attributes is now a static set (found in common/attributes.go). The API still takes a `map[string]string` but all public methods (`AddNode`, `AddAttr` etc) will call `FromStringMap()`, which will create a `Attrs` from the string map, and then pass them into private methods (`addNode`, `addAttr`, etc)

This way, the API stays the same, per #4 . 

Caveat: I have also silenced the `panic(err)` in almost all of them because the test cases themselves are wrong/have wrong attributes. I'll probably leave that to someone else to fix.  (grep for the below:)

``````
if err != nil {
    // panic(err)  // TODO: Fix...
}```
``````

### CODE REVIEW

**Comment by chewxy** [comment]
p/s: I know I promised typechecking for values as well, but it turns out I was reading the manual wrong, and some value types (like ArrowType) have their own grammars, which I was a bit too lazy to write.

File: escape_test.go [hunk 1]
```diff
@@ -35,8 +35,8 @@ func TestEscape(t *testing.T) {
 	if !strings.HasPrefix(s, `digraph "asdf adsf" {
 	"kasdf99 99"->7;
 	"a &lt;&lt; b";
-	"kasdf99 99" [ "<asfd"=1 ];
-	7 [ "<asfd"=1 ];
+	"kasdf99 99" [ color=green ];
```
Comment (awalterschulze) [inline escape_test.go]: This test testing escaping of string, so maybe you could change green to be something that needs escaping?

Comment (chewxy) [inline escape_test.go]: See latest file. Also, I note an error, but it's fine for now

**Comment by awalterschulze** [comment]
If I look at the two pull requests, basically we have to return a runtime error, which breaks the API or we have to make it type safe and get a compile error, which also breaks the API.

**Comment by awalterschulze** [comment]
Ok so the Attribute Type Descriptions are still not checked, hmmm.
That is definitely extra work that can be done separately.
But the decision about runtime vs compile time errors needs to made with that work in mind.

**Comment by chewxy** [comment]
Yeah, the attribute keys are static and any input string is checked against the static list. Values are still not typechecked.

**Comment by awalterschulze** [comment]
Hmmm yes that is probably true :)
Sorry I am taking so long with this.
I have not forgotten about htis.

**Comment by awalterschulze** [comment]
Ok first please move the common package to a package named attrs or attributes please.

Next anyone implementing the 
```
type Interface interface
```
would receive a compile error if we change that interface.

Ok what about making checking valid attribute names as a separate parsing step.
```
ParseBytes(dotBytes []byte) (*ast.Graph, error)
```
Already returns an error and it should be pretty easy to add a visitor step that simply checks that all attributes are valid?
This way at least string graphviz returns a parse error.

This leaves graphs that are created by the API user.
Those users that want strict checking should parse their output string.
We can add this to the readme.
Does that sound reasonable?

I think that is the only way without making a breaking change.

The other way is just to ask the one listed user if we can break the API.
In this case we should probably make a v0.1 release and include our new API as part of v0.2

**Comment by awalterschulze** [comment]
Sorry for the long wait, this is quite a big context switch for me.

**Comment by chewxy** [comment]
OK. I'll do that this weekend (sorry.. kinda busy week this week)

**Comment by awalterschulze** [comment]
No problem, sorry again for taking so long to reply.

**Comment by awalterschulze** [comment]
I have decided that we are going to make a breaking change to return some errors.
Or at least that is what this proposal is about
https://github.com/awalterschulze/gographviz/issues/23
Hopefully we can then fix this keyword problem as well.

**Comment by awalterschulze** [comment]
I hope this pull request can be replaced by this one
https://github.com/awalterschulze/gographviz/pull/25

**Comment by chewxy** [comment]
Oh holy crap I completely forgot about this one. Thanks !

**Comment by awalterschulze** [comment]
Maybe you would also like to review the change and would you like to be added to the AUTHORS or CONTRIBUTORS list?

**Comment by chewxy** [comment]
Looks good. [gorgonia](https://github.com/chewxy/gorgonia) will switch to using this version immediately. Also add my name to CONTRIB if you think I actually did contribute (I don't feel like I did)

**Comment by awalterschulze** [comment]
Would you mind if I add gorgonia to the user list as well?

**Comment by chewxy** [comment]
Yeah sure, go ahead. Thanks for the great update

**Comment by awalterschulze** [comment]
Could you provide an email and name for the CONTRIBUTORS list?

**Comment by chewxy** [comment]
Xuanyi Chew - chewxy [at] gmail.com

**Comment by awalterschulze** [comment]
Thanks both updated :)

**Pull request closed without merging**